var/spack/repos/builtin/packages/spectre/package.py

```diff
@@ -0,0 +1,189 @@
+    astrophysics and gravitational physics. In the future, we hope that it can
+    be applied to problems across discipline boundaries in fluid dynamics,
+    geoscience, plasma physics, nuclear physics, and engineering. It runs at
+    petascale and is designed for future exascale computers.
+
+    SpECTRE is being developed in support of our collaborative Simulating
+    eXtreme Spacetimes (SXS) research program into the multi-messenger
+    astrophysics of neutron star mergers, core-collapse supernovae, and
+    gamma-ray bursts."""
+
+    homepage = "https://spectre-code.com"
+    url = "https://github.com/sxs-collaboration/spectre/archive/v2021.12.15.tar.gz"
+    git = "https://github.com/sxs-collaboration/spectre.git"
+
+    maintainers = ['nilsleiffischer']
+
+    generator = 'Ninja'
+
+    version('develop', branch='develop')
+    version('2022.01.03', sha256='872a0d152c19864ad543ddcc585ce30baaad4185c2617c13463d780175cbde5f')
+    version('2021.12.15', sha256='4bfe9e27412e263ffdc6fcfcb84011f16d34a9fdd633ad7fc84a34c898f67e5c')
+
+    # Configuration variants
+    variant('build_type',
+            default='Release',
+            values=("Release", "Debug", "RelWithDebInfo"),
+            description='CMake build type')
+    variant('executables',
+            values=any_combination_of(
+                'SolvePoisson1D', 'SolvePoisson2D', 'SolvePoisson3D',
+                'SolveElasticity2D', 'SolveElasticity3D', 'SolveXcts',
+            ),
+            description="Executables to install")
+    variant('python', default=False, description="Build Python bindings")
+    variant('doc', default=False, description="Build documentation")
+    # TODO: support installation of executables with shared libs
+    # variant('shared',
+    #         default=False,
+    #         description="Build shared libraries instead of static")
+    variant('pch',
+            default=True,
+            description="Compile with pre-compiled headers")
+    variant('memory_allocator',
+            values=('system', 'jemalloc'),
+            multi=False,
+            default='system',
+            description="Which memory allocator to use")
+    variant('formaline',
+            default=True,
+            description=("Write the source tree into simulation output files "
+                         "to increase reproducibility of results"))
+    variant('profiling',
+            default=False,
+            description="Enable options to make profiling SpECTRE easier")
+
+    # Compiler support
+    conflicts('%gcc@:6')
+    conflicts('%clang@:7')
+    conflicts('%apple-clang@:10')
+
+    # Build dependencies
+    depends_on('cmake@3.12:3', type='build')
+    depends_on('ninja', type='build')
+    depends_on('python@2.7:3.10', type='build')
+
+    # Link dependencies
+    depends_on('charmpp@6.10.2:7.0')
+    depends_on('blaze@3.8')
+    depends_on('boost@1.60:1.77')
+    depends_on('brigand@master')
+    depends_on('gsl')
+    depends_on('hdf5@:1.12~mpi')
```

Comment:
Does the code break when an MPI-enabled HDF5 library is installed? If not, leave out the `~mpi`, so that Spack has more freedom to choose.

Reply:
My intention was to avoid unnecessary installs of MPI. Unless `charmpp` is built with MPI, we don't need it. I guess it's up to the user, but it feels wrong to build such an expensive dependency by default without really needing it. What do you think?

Reply:
Spack is often used on HPC system, and generally the `+mpi` variants are the default (e.g. in the `hdf5` package). You can't really change that choice of default in HDF5 in your package.

Note that your `~mpi` wouldn't set a default, it would stop anybody from using an MPI enabled HDF5 in a build that includes SpECTRE.

I would add a comment here explaining why you are using `~mpi`, and possibly open a bug report with HDF5 and/or libsharp. The behaviour you see looks like a bug. I definitively use HDF5 with MPI without any problems.

Reply:
My understanding is that you have to `find_package(MPI)` in CMake when `find_package(HDF5)` reports `HDF5_IS_PARALLEL` (see https://stackoverflow.com/questions/51561033/using-serial-hdf5-c-with-cmake). We don't currently do that in our build system because we don't use MPI, so I think I have to constrain `~mpi` in Spack, at least until we have added support to our build system upstream. Without finding and linking MPI in our build system, HDF5 will fail to include an MPI header. I added a comment to this effect.

Reply:
I fixed this upstream now, and removed the constraint on `hdf5`